changelog/13348.misc.rst

```diff
@@ -0,0 +1 @@
+Clean up type hints on :attr:`pytest.TestReport.when` and add a test.
```

Comment:
I think we can remove the changelog snippet here, as from the user's perspective this is one change.